File: concrete/src/main/java/pcb/uwu/amount/composite/electromagnetism/Coulombs.java
package pcb.uwu.amount.composite.electromagnetism;

import pcb.uwu.amount.composite.termodynamics.Joules;
import pcb.uwu.amount.quantity.Amperes;
import pcb.uwu.amount.quantity.ElectricCurrent;
import pcb.uwu.amount.quantity.Seconds;
import pcb.uwu.amount.quantity.Time;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.units.composite.electromagnetism.ElectricChargeUnit;

import java.math.BigDecimal;
import java.math.MathContext;

import static pcb.uwu.unit.composite.electromagnetism.CoulombUnit.COULOMB;
import static pcb.uwu.unit.composite.electromagnetism.FaradUnit.FARAD;
import static pcb.uwu.unit.composite.electromagnetism.VoltUnit.VOLT;
import static pcb.uwu.unit.quantity.AmpereUnit.AMPERE;
import static pcb.uwu.unit.quantity.SecondUnit.SECOND;

public class Coulombs extends ElectricCharge {

	// region constructors

	public Coulombs(Number value) {
		super(value, COULOMB);
	}

	public Coulombs(Number value, Magnitude magnitude) {
		super(value, magnitude, COULOMB);
	}

	public Coulombs(String value) {
		super(value, COULOMB);
	}

	public Coulombs(String value, Magnitude magnitude) {
		super(value, magnitude, COULOMB);
	}

	public Coulombs(BigDecimal value) {
		super(value, COULOMB);
	}

	public Coulombs(BigDecimal value, Magnitude magnitude) {
		super(value, magnitude, COULOMB);
	}

	public Coulombs(BigDecimalAmount amount) {
		super(amount, COULOMB);
	}

	public Coulombs(BigDecimalAmount amount, Magnitude magnitude) {
		super(amount, magnitude, COULOMB);
	}

	// endregion

	// region implement UnitAmount

	@Override
	public Coulombs plus(UnitAmount<ElectricChargeUnit> other, MathContext mathContext) {
		return new Coulombs(plusAmount(other, mathContext));
	}

	@Override
	public Coulombs minus(UnitAmount<ElectricChargeUnit> other, MathContext mathContext) {
		return new Coulombs(minusAmount(other, mathContext));
	}

	@Override
	public Coulombs multipliedBy(BigDecimal other, MathContext mathContext) {
		return new Coulombs(multipliedByScalar(other, mathContext));
	}

	@Override
	public Coulombs dividedBy(BigDecimal other, MathContext mathContext) {
		return new Coulombs(dividedByScalar(other, mathContext));
	}

	@Override
	public Coulombs convertTo(ElectricChargeUnit unit) {
		return new Coulombs(getAmountIn(unit));
	}

	// endregion

	// region composition

	public Amperes dividedBy(Time time, MathContext mathContext) {
		return new Amperes(getAmount().dividedBy(time.getAmountIn(SECOND), mathContext));
	}

	public Seconds dividedBy(ElectricCurrent electricCurrent, MathContext mathContext) {
		return new Seconds(getAmount().dividedBy(electricCurrent.getAmountIn(AMPERE), mathContext));
	}

	public Volts dividedBy(ElectricCapacitance electricCapacitance, MathContext mathContext) {
		return new Volts(getAmount().dividedBy(electricCapacitance.getAmountIn(FARAD), mathContext));
	}

	public Farads dividedBy(ElectricPotential electricPotential, MathContext mathContext) {
		return new Farads(getAmount().dividedBy(electricPotential.getAmountIn(VOLT), mathContext));
	}

	public Joules multipliedBy(ElectricPotential electricPotential, MathContext mathContext) {
		return new Joules(getAmount().multipliedBy(electricPotential.getAmountIn(VOLT), mathContext));
	}

	// endregion
}

File: implementation/src/main/java/pcb/uwu/amount/derived/electromagnetism/Coulombs.kt
package pcb.uwu.amount.derived.electromagnetism

import pcb.uwu.amount.base.Amperes
import pcb.uwu.amount.base.ElectricCurrent
import pcb.uwu.amount.base.Seconds
import pcb.uwu.amount.base.Time
import pcb.uwu.amount.derived.thermodynamics.Joules
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.AmpereUnit
import pcb.uwu.unit.base.SecondUnit
import pcb.uwu.unit.derived.electromagnetism.CoulombUnit
import pcb.uwu.unit.derived.electromagnetism.ElectricChargeUnit
import pcb.uwu.unit.derived.electromagnetism.FaradUnit
import pcb.uwu.unit.derived.electromagnetism.VoltUnit

class Coulombs : ElectricCharge
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, CoulombUnit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL)
            : super(amount, magnitude, CoulombUnit)

    // region UnitAmount

    override fun plus(electricCharge: UnitAmount<ElectricChargeUnit>) =
        Coulombs(this.amount + (electricCharge to CoulombUnit).amount)

    override fun minus(electricCharge: UnitAmount<ElectricChargeUnit>) =
        Coulombs(this.amount - (electricCharge to CoulombUnit).amount)

    override fun times(scalar: Number) =
        Coulombs(this.amount * scalar)

    override fun div(scalar: Number) =
        Coulombs(this.amount / scalar)

    // endregion

    // region composition

    operator fun div(time: Time) =
        Amperes(this.amount / (time to SecondUnit).amount)

    operator fun div(electricCurrent: ElectricCurrent) =
        Seconds(this.amount / (electricCurrent to AmpereUnit).amount)

    override fun div(electricCapacitance: ElectricCapacitance) =
        Volts(this.amount / (electricCapacitance to FaradUnit).amount)

    override fun div(electricPotential: ElectricPotential) =
        Farads(this.amount / (electricPotential to VoltUnit).amount)

    override fun times(electricPotential: ElectricPotential) =
        Joules(this.amount * (electricPotential to VoltUnit).amount)

    // endregion
}
